File: detection/semgrep/fixtures/csharp/authz-scope-from-request-header.cs
// Test fixture for rule: mcp-authz-scope-from-request-header-csharp
//
// S10 (forwarded-header-as-scope). ASP.NET Core MCP handler receives request headers.
// A handler that trusts a tenant/identity header (X-Org-Id, or X-Forwarded-For
// for IP scoping) as the authorization scope has the same client-controlled
// input as the JS sibling. The legitimate-neighbor reads a non-scope header
// (X-Request-Id) for logging and must stay silent.
//
// Corpus signatures (5 total, all POOL - TrashMobMCP, TrashMob):
// - TrashMobMCP/ApiKeyAuthMiddleware: context.Request.Headers.TryGetValue(ApiKeyHeaderName, out var headerValue) → auth
// - TrashMob/ChannelKeyAuthenticationFilter: context.HttpContext.Request.Headers.TryGetValue("IFTTT-Channel-Key", out var iftttChannelKeyRequest) → auth
// - TrashMob/WaiversV2Controller: Request.Headers["X-Forwarded-For"].ToString() → spoofable header in decision
// - TrashMob/CorrelationIdMiddleware: context.Request.Headers[CorrelationIdHeader].ToString() (SILENT: tracing)
// - TrashMob/PrivoApiKeyAuthenticationFilter: context.HttpContext.Request.Headers.TryGetValue(ApiKeyHeaderName, out var apiKeyHeader) → auth
//
// Two-way canary axis: is the header read (inbound, for a decision) or written (outbound) /
// read for a non-authz concern (locale, tracing, negotiation)? (write or non-authz ⇒ SILENT)

using Microsoft.AspNetCore.Http;
using System.Threading.Tasks;

// FIRE: Request.Headers["X-Forwarded-For"] / .TryGetValue(ApiKeyHeaderName, ...) used to authenticate/scope
public async Task<IEnumerable<Note>> NoteGetScopedVuln(HttpRequest request, string token) {
    var session = ResolveSession(token);
    // ruleid: mcp-authz-scope-from-request-header-csharp
    var headerOrg = request.Headers["x-org-id"].FirstOrDefault() ?? session.OrgId;
    return await _repo.Notes.Where(n => n.OrgId == headerOrg).ToListAsync();
}

// FIRE: TryGetValue with authz decision
public async Task<IEnumerable<Note>> NoteGetScopedTryGetValueVuln(HttpRequest request, string token) {
    var session = ResolveSession(token);
    // ruleid: mcp-authz-scope-from-request-header-csharp
    if (request.Headers.TryGetValue("X-Org-Id", out var headerOrgVal)) {
        var headerOrg = headerOrgVal.FirstOrDefault();
        return await _repo.Notes.Where(n => n.OrgId == headerOrg).ToListAsync();
    }
    return await _repo.Notes.Where(n => n.OrgId == session.OrgId).ToListAsync();
}

// FIRE (IP-scoping variant): X-Forwarded-For used to select client IP scope
public async Task<IEnumerable<Note>> NoteGetScopedByIpVuln(HttpRequest request, string token) {
    var session = ResolveSession(token);
    // ruleid: mcp-authz-scope-from-request-header-csharp
    var fwd = request.Headers["x-forwarded-for"].FirstOrDefault() ?? session.ClientIp;
    return await _repo.Notes.Where(n => n.ClientIp == fwd).ToListAsync();
}

// FIRE: header used directly as OrgId in a query filter (authz decision)
public async Task<Note> NoteGetByQueryVuln(HttpRequest request, string token, string id) {
    var session = ResolveSession(token);
    // ruleid: mcp-authz-scope-from-request-header-csharp
    var headerOrg = request.Headers["x-org-id"].FirstOrDefault() ?? session.OrgId;
    return await _repo.Notes.FirstOrDefaultAsync(n => n.Id == id && n.OrgId == headerOrg);
}

// OK: scope comes from the session; no header is consulted at all.
public async Task<IEnumerable<Note>> NoteGetScopedFixed(HttpRequest request, string token) {
    var session = ResolveSession(token);
    // ok: mcp-authz-scope-from-request-header-csharp
    return await _repo.Notes.Where(n => n.OrgId == session.OrgId).ToListAsync();
}

// OK (legitimate neighbor): a non-scope header read for logging only.
public void LogRequest(HttpRequest request) {
    // ok: mcp-authz-scope-from-request-header-csharp
    var reqId = request.Headers["x-request-id"].FirstOrDefault();
    _logger.LogInformation("tool invoked, reqId={ReqId}", reqId);
}

// SILENT: scope-shaped header (X-Forwarded-For) read ONLY for logging, no authz decision.
// This is the two-way canary: reading XFF for logging must NOT fire.
public void LogClientIp(HttpRequest request) {
    // ok: mcp-authz-scope-from-request-header-csharp
    var clientIp = request.Headers["x-forwarded-for"].FirstOrDefault();
    _logger.LogInformation("rate-limit check, clientIp={ClientIp}", clientIp);
}

// SILENT: scope-shaped header read into a variable but NEVER used for authz
public async Task<IEnumerable<Note>> ReadHeaderButDontUse(HttpRequest request, string token) {
    var session = ResolveSession(token);
    // ok: mcp-authz-scope-from-request-header-csharp
    var _unused = request.Headers["x-org-id"].FirstOrDefault();
    return await _repo.Notes.Where(n => n.OrgId == session.OrgId).ToListAsync();
}

// SILENT: outbound write on HttpRequestMessage — request.Headers.Add("Accept", ...) / request.Headers.Authorization = ...
public void OutboundWrite(System.Net.Http.HttpRequestMessage request) {
    // ok: mcp-authz-scope-from-request-header-csharp
    request.Headers.Add("Accept", "application/json");
    request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", "token");
}

// SILENT: header read for tracing (User-Agent, X-Correlation-ID)
public void LogTracing(HttpRequest request) {
    // ok: mcp-authz-scope-from-request-header-csharp
    var ua = request.Headers["User-Agent"].FirstOrDefault();
    var corr = request.Headers["X-Correlation-ID"].FirstOrDefault();
    _logger.LogInformation("ua={UA} corr={Corr}", ua, corr);
}

class Session { public string OrgId { get; set; } = ""; public string ClientIp { get; set; } = ""; }
class Note { public string Id { get; set; } = ""; public string OrgId { get; set; } = ""; public string ClientIp { get; set; } = ""; }
interface IRepository { 
    System.Collections.Generic.IEnumerable<Note> Notes { get; }
}
interface ILogger { void LogInformation(string message, params object[] args); }

Session ResolveSession(string token) => new Session();
IRepository _repo = null;
ILogger _logger = null;

File: detection/semgrep/fixtures/kotlin/authz-scope-from-request-header.kt
// Test fixture for rule: mcp-authz-scope-from-request-header-kotlin
//
// S10 (forwarded-header-as-scope). Ktor/Spring MCP handler receives request headers.
// A handler that trusts a tenant/identity header (X-Org-Id, or X-Forwarded-For
// for IP scoping) as the authorization scope has the same client-controlled
// input as the JS sibling. The legitimate-neighbor reads a non-scope header
// (X-Request-Id) for logging and must stay silent.
//
// Corpus signatures (6 total, all POOL — STRONGEST CELL, MCP-NATIVE):
// - modelcontextprotocol/kotlin-sdk KtorServer.kt: val sessionId = call.request.header(MCP_SESSION_ID_HEADER) → transportManager.getTransport(sessionId) — CANONICAL S10
// - modelcontextprotocol/kotlin-sdk HostValidation.kt: val hostHeader = call.request.header(HttpHeaders.Host) → host-validation decision
// - modelcontextprotocol/kotlin-sdk StreamableHttpServerTransport.kt: val sessionId = call.request.header(MCP_SESSION_ID_HEADER) → session lookup
// - PortSwigger/mcp-server KtorServerManager.kt: val origin = call.request.header("Origin"); val host = call.request.header("Host") → DNS-rebinding / origin guard
// - six2dez/burp-ai-agent McpAccessControlPlugin.kt: RequestFacts(origin = call.request.headers["Origin"], host = call.request.headers["Host"]) → access control
// - six2dez/burp-ai-agent KtorMcpServerManager.kt: val authHeader = call.request.header(...) → auth
//
// Ecosystem corroboration (Ktor auth interceptors):
// - csieflyman/multi-projects: val apiKey = call.request.header(AuthConst.API_KEY_HEADER_NAME) in onAuthenticate
// - f-arslan/GptMap: val idToken = call.request.header("Authorization")?.removePrefix("Bearer ")
// - Elfocrash/L2jRest: val headerApiKey = context.call.request.header("x-api-key") → auth compare
//
// Two-way canary axis: header read for auth/session-scope/validation ⇒ FIRE; header read for tracing/negotiation ⇒ SILENT

import io.ktor.server.request.*
import io.ktor.http.*

// FIRE: call.request.header(MCP_SESSION_ID_HEADER) → getTransport(...) — CANONICAL S10
suspend fun noteGetScopedCanonicalVuln(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val sessionId = call.request.header("Mcp-Session-Id")
    return transportManager.getTransport(sessionId)
}

// FIRE: call.request.header("Authorization"|"x-api-key"|"Origin"|"Host") → auth/validation decision
suspend fun noteGetScopedAuthVuln(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val authHeader = call.request.header("Authorization")?.removePrefix("Bearer ")
    return repo.authenticate(authHeader)
}

// FIRE: header value used to select org scope in a repository call (authz decision)
suspend fun noteGetScopedVuln(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val headerOrg = call.request.headers["x-org-id"] ?: session.orgId
    return repo.listNotesByOrg(headerOrg)
}

// FIRE: call.request.header() form with authz decision
suspend fun noteGetScopedHeaderVuln(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val headerOrg = call.request.header("x-org-id") ?: session.orgId
    return repo.listNotesByOrg(headerOrg)
}

// FIRE (IP-scoping variant): X-Forwarded-For used to select client IP scope
suspend fun noteGetScopedByIpVuln(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val fwd = call.request.headers["x-forwarded-for"] ?: session.clientIp
    return repo.listNotesForClientIp(fwd)
}

// FIRE: header used directly as orgId in a query filter (authz decision)
suspend fun noteGetByQueryVuln(call: ApplicationCall, token: String, id: String) {
    val session = resolveSession(token)
    // ruleid: mcp-authz-scope-from-request-header-kotlin
    val headerOrg = call.request.headers["x-org-id"] ?: session.orgId
    return repo.findNoteBy(id = id, orgId = headerOrg)
}

// OK: scope comes from the session; no header is consulted at all.
suspend fun noteGetScopedFixed(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ok: mcp-authz-scope-from-request-header-kotlin
    return repo.listNotesByOrg(session.orgId)
}

// OK (legitimate neighbor): a non-scope header read for logging only.
suspend fun logRequest(call: ApplicationCall) {
    // ok: mcp-authz-scope-from-request-header-kotlin
    val reqId = call.request.headers["x-request-id"]
    logger.info { "tool invoked, reqId=$reqId" }
}

// SILENT: scope-shaped header (X-Forwarded-For) read ONLY for logging, no authz decision.
// This is the two-way canary: reading XFF for logging must NOT fire.
suspend fun logClientIp(call: ApplicationCall) {
    // ok: mcp-authz-scope-from-request-header-kotlin
    val clientIp = call.request.headers["x-forwarded-for"]
    logger.info { "rate-limit check, clientIp=$clientIp" }
}

// SILENT: scope-shaped header read into a variable but NEVER used for authz
suspend fun readHeaderButDontUse(call: ApplicationCall, token: String) {
    val session = resolveSession(token)
    // ok: mcp-authz-scope-from-request-header-kotlin
    val _unused = call.request.headers["x-org-id"]
    return repo.listNotesByOrg(session.orgId)
}

// SILENT: header read for tracing/negotiation (X-Real-IP, User-Agent, Content-Type, X-GitHub-Event)
suspend fun logTracing(call: ApplicationCall) {
    // ok: mcp-authz-scope-from-request-header-kotlin
    val realIp = call.request.header("X-Real-IP")
    val ua = call.request.headers["User-Agent"]
    val event = call.request.header("X-GitHub-Event")
    logger.info { "tracing realIp=$realIp ua=$ua event=$event" }
}

fun resolveSession(token: String): Session = TODO()
class Session(val orgId: String, val clientIp: String)
class Repo { 
    fun listNotesByOrg(orgId: String) = TODO()
    fun listNotesForClientIp(ip: String) = TODO()
    fun findNoteBy(id: String, orgId: String) = TODO()
    fun authenticate(token: String?) = TODO()
}
val repo = Repo()
val transportManager = object { fun getTransport(id: String?) = TODO() }
val logger = TODO()